### DISCUSSION TITLE
The pipeline always run on a bad agent.

### DISCUSSION THREAD

**Discussion by zszen** [gocd/gocd#12581] [Q&A] [answered]
I have two agent for gocd-server. One on ubuntu installed with apt tool. Another run on docker compose (it is not good). 

I duplicate a pipeline. It run on the bad agent. And it don't change agent after I remove the docker-agent.  I also input the resource keywords for lock on the right one. But it don't work. .

Now I have only one agent. And the pipeline always waiting a agent. ;(

![](https://github.com/gocd/gocd/assets/1890538/de51e195-3943-4bcf-abe5-43f8ff7bb1f5)

**Comment by chadlwilson** [comment]
Can you share your agents page? And the job configuration for the `def` job in that pipeline (I.e which resources it requires to run)

**Reply by zszen** [reply]
pipeline job's resource
![](https://github.com/gocd/gocd/assets/1890538/7fda433e-6c47-4c5b-ac66-886c1b2ea0db)

agents list
![](https://github.com/gocd/gocd/assets/1890538/327c80f7-7752-4ee2-b002-ce7e968d7781)

history. The task timeout
![](https://github.com/gocd/gocd/assets/1890538/618eeb01-1909-4711-b2ae-dc038ce5649e)

**Reply by chadlwilson** [reply]
Is your 'nacos_standalone' pipeline assigned to the 'docker' environment?

Your agent is configured with that environment so will only run matching pipelines.

If not, you can either remove the environment from the agent or map the pipeline.

**Reply by zszen** [reply, marked as answer]
I deleted the problem agent yet before that. the only agent is the right env for pipeline. I have two pipelines. The first is correct runing on the ubuntu agent. But the problem one can't re-choice agent. Always waiting for the problem agent which I deleted.

I found that I add nacos pipeline to env. it works. I don't understand. does not global env only set env variable only ?

![](https://github.com/gocd/gocd/assets/1890538/13d2c5e8-71e6-49d2-aa68-6390646ecb70)

**Reply by chadlwilson** [reply]
Environments are a way to both set environment variables for ALL pipelines, as well as a way to segregate where ALL jobs in a pipeline run, without setting resources at the job level.

For example, if an agent used for deployment only has access to the production environment, you want to reflect that so agents for other environments don't pick up jobs for production.

You don't need to use environments if you don't want to use either of these capabilities. Agents _without_ an environment set will pick up jobs for pipelines which ALSO don't have an environment (aka the "default“ environment)

Use the docs :)
https://docs.gocd.org/current/configuration/managing_environments.html
